src/thing_creature.c

```diff
@@ -5364,6 +5369,481 @@ void illuminate_creature(struct Thing *creatng)
+                        {
+                            if (room->used_capacity < room->total_capacity)
+                            {
+                                make_creature_conscious(droptng);
+                                initialise_thing_state(droptng, CrSt_CreatureArrivedAtPrison);
+                                struct CreatureControl* dropctrl = creature_control_get_from_thing(droptng);
+                                dropctrl->flgfield_1 |= CCFlg_NoCompControl;
+                            }
+                            else
+                            {
+                               output_message(SMsg_PrisonTooSmall, 0, true); 
+                            }
+                        }
+                    }
+                    break;
+                }
+            }
+        }
+    }
+}
+
+void direct_control_pick_up_or_drop(struct PlayerInfo *player)
+{
+    struct Thing *thing = thing_get(player->controlled_thing_idx);
+    struct CreatureControl* cctrl = creature_control_get_from_thing(thing);
+    struct Thing* dragtng = thing_get(cctrl->dragtng_idx);
+    if (!thing_is_invalid(dragtng))
+    {
+        if (thing_is_trap_crate(dragtng))
+        {
+            struct Thing *traptng = thing_get(player->thing_under_hand);
+            if (!thing_is_invalid(traptng))
+            {
+                if (traptng->class_id == TCls_Trap)
+                {   
+                    cctrl->arming_thing_id = traptng->index;
+                    internal_set_thing_state(thing, CrSt_CreatureArmsTrap);
+                    return;
+                }
+            }
+        }
+        controlled_creature_drop_thing(thing, dragtng);
+    }
+    else
+    {
+        struct Thing* picktng = thing_get(player->thing_under_hand);
+        struct Room* room;
+        if (!thing_is_invalid(picktng))
+        {
+            if (object_is_gold_pile(picktng))
+            {
+                struct CreatureStats* crstat = creature_stats_get_from_thing(thing);
+                if (thing->creature.gold_carried < crstat->gold_hold)
+                {
+                    cctrl->pickup_object_id = picktng->index;
+                    internal_set_thing_state(thing, CrSt_ImpPicksUpGoldPile);
+                    return;
+                }
+                else
+                {
+                    if (is_thing_directly_controlled_by_player(thing, my_player_number))
+                    {
+                        play_non_3d_sample(119);
+                        return;
+                    }
+                }
+            }
+            room = get_room_thing_is_on(picktng);
+            if (!room_is_invalid(room))
+            {
+                if ( (room->kind == RoK_WORKSHOP) && (room->owner == thing->owner) )
+                {
+                    if (thing_is_workshop_crate(picktng))
+                    {
+                        if (!imp_will_soon_be_getting_object(thing->owner, picktng))
```

Comment:
What does this do here, can't we pick up objects already considered by another imp? If so, that does not seem like a smart idea.

Reply:
Will it cause problems if you pick something another Imp was going to?

Reply:
No. You can do this outside of possession too, if an imp is on the way, assign a new imp to more quickly pick it up. The other imp will just 'deal with it' if he gets to the workshop and find nothing.